File: Binary Tree Preorder Traversal/Binary_Tree_Preorder_Traversal.cpp
// 递归的过于简单了，这里就直接使用堆栈来做了
// 0ms 100%

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<vector>
#include<stack>
using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> values;
        stack<TreeNode*> treeStack;
        if (!root)        
            return values;
        
        while (root || !treeStack.empty())
        {
            while (root)
            {
                values.push_back(root->val);
                treeStack.push(root);
                root = root->left;
            }

            if (!treeStack.empty())
            {
                root = treeStack.top();
                treeStack.pop();

                root = root->right;
            }
        }
        return values;
    }
};

File: Binary Tree Preorder Traversal/Binary_Tree_Preorder_Traversal.py
# 52ms 99.00%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        values = []
        treeStack = []

        if (root is None):
            return values
        while root or len(treeStack) > 0:
            while root:
                values.append(root.val)
                treeStack.append(root)
                root = root.left
            if len(treeStack) > 0:
                root = treeStack.pop()
                root = root.right
        return values